File: Queue/queue.py
#===============================================================================
# implementation of Queue and Priority Queue using Arrays
#===============================================================================
import numpy
import random

class Queue:
    
#===============================================================================
#     Arrays, linked lists, trees, etc. are best for
#     data that represents real objects.
# 
#     Stacks & Queues are instead used to complete a 
#     task and are soon after discarded.
# 
#     Stacks & Queues
#     1. Allow only a single item to be added or removed at a time
#     2.  Queues allow access to the first item inserted (FIFO)
#===============================================================================
    
    #===========================================================================
    # initializing Queue with specific size
    #===========================================================================
    def __init__(self, queue_size):
        self.queue_size = queue_size
        self.queue_array = numpy.array([None] * queue_size)

        #Sets stack as empty
        self.front , self.rear, self.number_of_items = 0, 0, 0
        
        print "Queue of size %d initialized." %(self.queue_size)
        self.display_queue()
        
    def display_queue(self):
        print "[ " ,
        for element in self.queue_array:
            if element != None:
                if self.queue_array[self.rear - 1] == element and self.queue_array[self.front] == element:
                    print element, " [F] [R], " ,
                elif self.queue_array[self.front] == element:
                    print element, " [F], " ,
                elif self.queue_array[self.rear - 1] == element:
                    print element, " [R], " ,
                else:
                    print element, " ," ,
            else:
                print " ," ,
        print " ]"
        
    def insert(self, value):
        if self.number_of_items + 1 <= self.queue_size:
            self.queue_array[self.rear] = value
            self.rear += 1
            self.number_of_items += 1
            print "INSERT : %s added in Queue" %(value)
            self.display_queue()
        else:
            print "Sorry but Queue is already full."
            self.display_queue()
            
    def insert_many(self, multiple_value):
        for value in multiple_value:
            self.insert(value)

    #===========================================================================
    # This priority insert will add items in order from high to low
    #===========================================================================
    def priority_insert(self, value):
        if ( self.number_of_items == 0):
            self.insert(value)
        else:
            for i in range(self.number_of_items - 1, -1, -1):
                if( value > self.queue_array[i]):
                    self.queue_array[i + 1] = self.queue_array[i]
                else: break
            self.queue_array[self.front] = value
            self.rear += 1
            self.number_of_items += 1
        self.display_queue()            

    def priority_insert_many(self, multiple_value):
        for value in multiple_value:
            self.priority_insert(value)
    
    def remove(self):
        if self.number_of_items > 0:
            print "REMOVE : value %s was removed from Queue" %(self.queue_array[self.front])
            self.queue_array[self.front] = None
            self.front += 1
            self.number_of_items -= 1
            self.display_queue()
        else:
            print "Queue is already empty"
            self.display_queue()
    
    def peek(self):
        if self.number_of_items > 0:
            print " %s is the first element in Queue" %(self.queue_array[self.front])
        else:
            print "Queue is already empty"
            

queue_size = int(raw_input("Enter the Queue size : "))
new_queue = Queue(queue_size)
new_queue.insert_many(random.sample(xrange(0, 100), queue_size))
new_queue.remove()
new_queue.peek()
#new_queue.priority_insert_many(random.sample(xrange(0, 100), queue_size))


            

File: java/Queue/Queue.java
/*implementation of Queue and Priority Queue using Arrays*/

/*
 	 Arrays, linked lists, trees, etc. are best for
     data that represents real objects.
 
     Stacks & Queues are instead used to complete a 
     task and are soon after discarded.
 
     Stacks & Queues
     1. Allow only a single item to be added or removed at a time
     2.  Queues allow access to the first item inserted (FIFO)
*/

public class Queue {
	
	private String[] queueArray;
	
	private int queueSize;
	
	//initializing queue as empty
	private int front, rear, numberOfItems = 0;
	
	public Queue(int size){
		queueSize = size;
		queueArray = new String[queueSize];
	}
	
	public void insert(String value){
		if (numberOfItems + 1 <= queueSize){
			queueArray[rear] = value;
			rear += 1;
			numberOfItems += 1;
			System.out.println(String.format("Insert : value %s was added to queue", value));
		} else{
			System.out.println("Sorry Queue is full.");
		}
	}
	
	public void insert_many(String[] values){
		for (String value : values){
			insert(value);
		}
	}
	
	public void remove(){
		if( numberOfItems >0 ){
			String value = queueArray[front];
			queueArray[front] = null;
			front += 1;
			numberOfItems -= 1;
			System.out.println(String.format("Remove : value %s was removed from queue", value));
		} else {
			System.out.println("Sorry Queue is already empty");
		}
	}
	
	public void peek(){
		if (numberOfItems > 0 ){
			System.out.println(String.format("Peek : value %s is in front of queue", queueArray[front]));
		} else {
			System.out.println("Sorry Queue is already empty");
		}
	}
	
//  This priority insert will add items in order from high to low
	public void priorityInsert(String input){
		int i;
		if(numberOfItems == 0){
			insert(input);
		} else {
			for(i = numberOfItems-1; i >= 0; i--){
				if(Integer.parseInt(input) > Integer.parseInt(queueArray[i])){
					queueArray[i+1] = queueArray[i];
				} else break; // Done shifting items in queue
			}
			queueArray[i+1] = input;
			rear += 1;
			numberOfItems += 1;
		}
	}
	
	public void priorityInsertMany(String[] values){
		for (String value : values){
			priorityInsert(value);
		}
	}

	public void displayQueue(){
		System.out.print("[ ");
		for( String value : queueArray){
			if (value != null){
				if (value == queueArray[front] && value == queueArray[rear]) {
					System.out.print(value + " [F] [R],");
				} else if ( value == queueArray[front]){
					System.out.print(value + " [F],");
				} else if (value == queueArray[rear - 1]){
					System.out.print(value + " [R],");
				} else {
					System.out.print(value + " ,");
				}
			} else{
				System.out.print(" ,");
			}
		}
		System.out.print(" ]\n");
	}
	
	public static void main(String [] args){		
		Queue theQueue = new Queue(10);
		theQueue.priorityInsert("16");
		theQueue.priorityInsert("25");
		theQueue.priorityInsert("10");
		theQueue.displayQueue();
		theQueue.remove();
		theQueue.peek();
		theQueue.displayQueue();		
	}
}